flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayExceptFunction.java

```diff
@@ -0,0 +1,134 @@
+import java.util.Map;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#ARRAY_EXCEPT}. */
+@Internal
+public class ArrayExceptFunction extends BuiltInScalarFunction {
+    private final ArrayData.ElementGetter elementGetter;
+    private final EqualityAndHashcodeProvider equalityAndHashcodeProvider;
+
+    public ArrayExceptFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.ARRAY_EXCEPT, context);
+        final DataType dataType =
+                ((CollectionDataType) context.getCallContext().getArgumentDataTypes().get(0))
+                        .getElementDataType()
+                        .toInternal();
+        elementGetter = ArrayData.createElementGetter(dataType.toInternal().getLogicalType());
+        this.equalityAndHashcodeProvider = new EqualityAndHashcodeProvider(context, dataType);
+    }
+
+    @Override
+    public void open(FunctionContext context) throws Exception {
+        equalityAndHashcodeProvider.open(context);
+    }
+
+    public @Nullable ArrayData eval(ArrayData arrayOne, ArrayData arrayTwo) {
+        try {
+            if (arrayOne == null || arrayTwo == null) {
+                return null;
+            }
+
+            List<Object> list = new ArrayList<>();
+            Map<ObjectContainer, Integer> map = new HashMap<>();
+            for (int pos = 0; pos < arrayOne.size(); pos++) {
+                final Object element = elementGetter.getElementOrNull(arrayOne, pos);
+                if (element == null) {
+                    map.put(null, map.getOrDefault(null, 0) + 1);
+                } else {
+                    ObjectContainer objectContainer = new ObjectContainer(element);
+                    map.put(objectContainer, map.getOrDefault(objectContainer, 0) + 1);
+                }
+            }
+            for (int pos = 0; pos < arrayTwo.size(); pos++) {
+                final Object element = elementGetter.getElementOrNull(arrayTwo, pos);
+                if (element == null) {
+                    if (map.containsKey(null)) {
+                        int count = map.get(null);
+                        count--;
+                        if (count == 0) {
+                            map.remove(null);
+                        } else {
+                            map.put(null, count);
+                        }
+                    }
+                } else {
+                    ObjectContainer objectContainer = new ObjectContainer(element);
+                    if (map.containsKey(objectContainer)) {
+                        int count = map.get(objectContainer);
+                        count--;
+                        if (count == 0) {
+                            map.remove(objectContainer);
+                        } else {
+                            map.put(objectContainer, count);
+                        }
+                    }
+                }
+            }
+            for (int pos = 0; pos < arrayOne.size(); pos++) {
```

Comment:
Maybe I am tired while reviewing the PR, but why do you go twice over `arrayOne`?

Shouldn't you build cardinality map from `arrayTwo` and when going throug `arrayOne` check if the element is present in the cardinality map. If the element is present don't add it to the list and decrease the count in the map, if it is no longer in the cardinality map add it to the list?

Reply:
Do you want to group all equal elements together? I feel it's a weird order guarantee.

Reply:
The purpose of iterating twice over arrayOne is to guarantee the order, since the order of keys in a map is not fixed.

Reply:
> Maybe I am tired while reviewing the PR, but why do you go twice over `arrayOne`?
> 
> Shouldn't you build cardinality map from `arrayTwo` and when going throug `arrayOne` check if the element is present in the cardinality map. If the element is present don't add it to the list and decrease the count in the map, if it is no longer in the cardinality map add it to the list?


If we build a cardinality map from arrayTwo, we still need to iterate over arrayOne twice. The first iteration is to decrease the count in the map, and the second is to guarantee the order. I think I can use LinkedHashMap, which can guarantee the order of the keys.